File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def is_valid(node: Optional[TreeNode], limit: int, is_left: bool) -> bool:
            # Return true if current node is empty because we have nothing to check
            if not node:
                return True

            # Compare differently based on the bool flag
            valid = node.val < limit if is_left else node.val > limit

            # Return false if invalid
            if not valid:
                return False

            return is_valid(node.left, limit, is_left) and is_valid(
                node.right, limit, is_left
            )

        def dfs(node: Optional[TreeNode]) -> bool:
            # Return true if current node is empty because we have nothing to check
            if not node:
                return True

            # Return false if any of the recursive call returns false
            if not is_valid(node.left, node.val, True) or not (
                is_valid(node.right, node.val, False)
            ):
                return False
            return dfs(node.left) and dfs(node.right)

        return dfs(root)

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        auto isValid = [&](this auto&& self, TreeNode* node, const int limit,
                           const bool isLeft) -> bool {
            // Return true if current node is empty because we have nothing to
            // check
            if (node == nullptr) {
                return true;
            }

            // Compare differently based on the bool flag
            bool valid = isLeft ? node->val < limit : node->val > limit;

            // Return false if invalid
            if (!valid) {
                return false;
            }
            return self(node->left, limit, isLeft) &&
                   self(node->right, limit, isLeft);
        };
        auto dfs = [&](this auto&& self, TreeNode* node) {
            // Return true if current node is empty because we have nothing to
            // check
            if (node == nullptr) {
                return true;
            }
            // Return false if any of the recursive call returns false
            if (!isValid(node->left, node->val, true) ||
                !isValid(node->right, node->val, false)) {
                return false;
            }
            return self(node->left) && self(node->right);
        };
        return dfs(root);
    }
};